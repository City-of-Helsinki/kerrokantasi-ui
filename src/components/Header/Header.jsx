/* eslint-disable import/no-unresolved */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { IconUser, IconSignin, Header as HDSHeader, Logo, LoadingSpinner } from 'hds-react';
import { FormattedMessage, useIntl } from 'react-intl';
import { connect, useDispatch } from 'react-redux';
import logoBlack from '@city-images/logo-fi-black.svg';
import logoSwedishBlack from '@city-images/logo-sv-black.svg';
import { NavLink, useLocation, useSearchParams } from 'react-router-dom';

import config from '../../config';
import getUser from '../../selectors/user';
import useAuthHook from '../../hooks/useAuth';
import { addToast } from '../../actions/toast';
import { createLocalizedNotificationPayload, NOTIFICATION_TYPES } from '../../utils/notify';
import { setLanguage } from '../../actions';

const Header = ({ user, onChangeLanguage }) => {
  const { authenticated, login, logout } = useAuthHook();
  const location = useLocation();
  const dispatch = useDispatch();
  const intl = useIntl();
  const [searchParams, setSearchParams] = useSearchParams();

  const language = intl.locale;
  
  const setLocale = (newLanguage) => {
    if (newLanguage !== language) {
      onChangeLanguage(newLanguage);
      dispatch(setLanguage(newLanguage));
    }
  };

  const doLogin = async () => {
    if (config.maintenanceDisableLogin) {
      dispatch(addToast(createLocalizedNotificationPayload(NOTIFICATION_TYPES.error, 'maintenanceNotificationText')));
      return;
    }
    if (!authenticated) {
      try {
        await login();
      } catch {
        dispatch(addToast(createLocalizedNotificationPayload(NOTIFICATION_TYPES.error, 'loginAttemptFailed')));
      }
    } else {
      logout();
    }
  };

  const onLanguageChange = (newLanguage) => {
    setLocale(newLanguage);
    const urlSearchParams = new URLSearchParams(window.location.search);

    urlSearchParams.set('lang', newLanguage);

    setSearchParams(urlSearchParams);
  };

  useEffect(() => {
    const langParam = searchParams.get('lang');

    if (langParam && langParam !== language) {
      setLocale(langParam);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchParams]);

  const getNavItem = (id, url, addSuffix = true) => {
    const active = location.pathname === url;
    let messageId = id;
    if (id === 'ownHearings' && (!user || user.adminOrganizations.length === 0)) {
      return null;
    }
    if (id === 'userInfo' && !user) {
      return null;
    }
    if (addSuffix) {
      messageId += 'HeaderText';
    }

    return (
      <FormattedMessage id={messageId} key={messageId}>
        {url
          ? (text) => <HDSHeader.Link as={<NavLink />} to={`${url}?lang=${language}`} label={text} active={active} />
          : (text) => <p>{text}</p>}
      </FormattedMessage>
    );
  };

  const languages = [
    { label: 'Suomi', value: 'fi', isPrimary: true },
    { label: 'Svenska', value: 'sv', isPrimary: true },
    { label: 'English', value: 'en', isPrimary: true },
  ];

  const logo = (
    <FormattedMessage id='headerLogoAlt'>
      {(altText) => <Logo src={language === 'sv' ? logoSwedishBlack : logoBlack} alt={altText} />}
    </FormattedMessage>
  );

  const navigationItems = [
    getNavItem('hearings', '/hearings/list'),
    getNavItem('hearingMap', '/hearings/map'),
    getNavItem('info', '/info'),
    getNavItem('ownHearings', '/user-hearings', false),
    getNavItem('userInfo', '/user-profile', false),
  ];

  return (
    <HDSHeader onDidChangeLanguage={onLanguageChange} languages={languages} defaultLanguage={language} key={language}>
      <HDSHeader.ActionBar
        title='Kerrokantasi'
        titleAriaLabel='Kerrokantasi'
        frontPageLabel='Kerrokantasi'
        titleHref='/'
        logoHref='/'
        openFrontPageLinksAriaLabel={<FormattedMessage id='headerOpenFrontPageLinks' />}
        logo={logo}
      >
        <HDSHeader.LanguageSelector />
        <HDSHeader.ActionBarItem
          fixedRightPosition
          label={user ? user.displayName : <FormattedMessage id='login' />}
          aria-label={user ? user.displayName : intl.formatMessage({ id: 'login' })}
          icon={user ? <IconUser /> : <IconSignin />}
          closeIcon={user ? <IconUser /> : <IconSignin />}
          closeLabel={user?.displayName}
          onClick={user ? () => {} : doLogin}
          id='action-bar-login'
          className={user ? 'logout-button' : 'login-button'}
        >
          {user && (
            <HDSHeader.ActionBarItem
              label={<FormattedMessage id='logout' />}
              closeLabel=''
              closeIcon={<LoadingSpinner small />}
              onClick={doLogin}
              id='action-bar-login'
              className='logout-button'
            />
          )}
        </HDSHeader.ActionBarItem>
      </HDSHeader.ActionBar>
      <HDSHeader.NavigationMenu>{navigationItems}</HDSHeader.NavigationMenu>
    </HDSHeader>
  );
};

Header.propTypes = {
  user: PropTypes.object,
  onChangeLanguage: PropTypes.func,
};

export default connect((state) => ({
  user: getUser(state),
}))(Header);
