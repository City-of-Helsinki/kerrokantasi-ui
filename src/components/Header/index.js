/* eslint-disable react/no-multi-comp */
import React from 'react';
import PropTypes from 'prop-types';
import { Navbar, Button, DropdownButton, MenuItem } from 'react-bootstrap';
import Icon from '../../utils/Icon';

import LanguageSwitcher from './LanguageSwitcher';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { LinkContainer } from 'react-router-bootstrap';
import { withRouter } from 'react-router-dom';
import Link from '../../components/LinkWithLang';
import throttle from 'lodash/throttle';
import scrolltop from 'scrolltop';
import {getUser} from '../../selectors/user';
import userManager from "../../utils/userManager";
import { toggleContrast } from "../../actions";

// eslint-disable-next-line import/no-unresolved
import logoBlack from '@city-images/logo-fi-black.svg';

// eslint-disable-next-line import/no-unresolved
import logoSwedishBlack from '@city-images/logo-sv-black.svg';
import config from "../../config";
import { localizedNotifyError } from '../../utils/notify';

class Header extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      navbarExpanded: false,
    };
  }
  toggleNavbar() {
    this.setState({navbarExpanded: !this.state.navbarExpanded});
  }

  componentDidMount() {
    if (typeof window !== 'undefined') {
      this._handleNavFix = throttle(() => {
        const scrollY = scrolltop();
        if (scrollY > 115) {
          document.body.classList.add('nav-fixed');
        } else {
          document.body.classList.remove('nav-fixed');
        }
      }, 25);
      window.addEventListener('scroll', this._handleNavFix, false);
    }
  }

  componentWillUnmount() {
    if (this._handleNavFix) {
      window.removeEventListener('scroll', this._handleNavFix);
    }
  }

  async handleLogin() {
    try {
      await userManager.signinRedirect({ ui_locales: this.props.language });
    } catch (error) {
      localizedNotifyError("loginAttemptFailed");
    }
  }

  getUserItems() {
    const {user} = this.props;

    if (user) {
      return [
        <DropdownButton
          pullRight
          key="profile"
          id="userMenu"
          className="user-menu user-menu--logged"
          title={
            <span>
              <Icon name="user" className="user-nav-icon" aria-hidden="true" />
              <span className="user-name">{user.displayName}</span>
            </span>
          }
        >
          <MenuItem
            key="logout"
            eventKey="logout"
            onClick={() => userManager.signoutRedirect()}
          >
            <FormattedMessage id="logout" />
          </MenuItem>
        </DropdownButton>,
      ];
    }
    return [
      <FormattedMessage key="login" id="login">
        {login => (
          <Button
            key="login"
            aria-label={login}
            className="user-menu login-link user-menu--unlogged"
            onClick={() => this.handleLogin()}
          >
            <Icon name="user-o" className="user-nav-icon" aria-hidden="true" />
            <span className="user-name">{login}</span>
          </Button>
        )}
      </FormattedMessage>
    ];
  }

  getNavItem(id, url, addSuffix = true) {
    const {history, language, user} = this.props;
    const active = history && history.location.pathname === url;
    let messageId = id;
    if (id === 'ownHearings' && (!user || user.adminOrganizations.length === 0)) {
      return null;
    }
    if (addSuffix) { messageId += 'HeaderText'; }
    const navLink = (
      <a href="#">
        <FormattedMessage id={messageId} />
      </a>
    );
    if (url) {
      // Can't use custom link component here because it will break the navigation
      // so LinkContainer must contain same logic
      return (
        <li className={`nav-item ${active ? 'active' : ''}`}>
          <LinkContainer to={url + '?lang=' + language} className="nav-link">
            {navLink}
          </LinkContainer>
        </li>
      );
    }
    return (
      <li className={`nav-item ${active && 'active'}`}>
        {navLink}
      </li>
    );
  }

  contrastToggle() {
    if (config.enableHighContrast) {
      return (
        <FormattedMessage key="contrastTitle" id="contrastTitle">
          {text => (
            <Button
              key="text"
              aria-label={text}
              className="contrast-button"
              onClick={() => this.props.toggleContrast()}
            >
              <Icon name="adjust"/>
              <span className="contrast-title">{text}</span>
            </Button>
          )}
        </FormattedMessage>
      );
    }
    return (
      <div />
    );
  }

  render() {
    const {language} = this.props;
    const userItems = this.getUserItems();
    return (
      <div>
        <FormattedMessage key="headerUserNavLabel" id="headerUserNavLabel">
          {headerUserNavLabel => (
            <Navbar fluid staticTop defaultExpanded className="navbar-kerrokantasi" aria-label={headerUserNavLabel}>
              <Navbar.Header>
                <Navbar.Brand>
                  <Link to={{ path: "/" }}>
                    <FormattedMessage id="headerLogoAlt">
                      {altText => <img
                        src={language === 'sv' ? logoSwedishBlack : logoBlack}
                        className="navbar-logo"
                        alt={altText}
                      />}
                    </FormattedMessage>
                  </Link>
                </Navbar.Brand>
              </Navbar.Header>

              <div className="nav-user-menu navbar-right">
                {this.contrastToggle()}
                <LanguageSwitcher currentLanguage={this.props.language}/>
                {userItems}
              </div>
            </Navbar>
          )}
        </FormattedMessage>

        <FormattedMessage id="headerPagesNavLabel">
          {headerPagesNavLabel => (
            <Navbar default fluid collapseOnSelect className="navbar-primary" aria-label={headerPagesNavLabel} onToggle={this.toggleNavbar.bind(this)}>
              <Navbar.Header>
                <Navbar.Brand>
                  <Link to={{path: "/"}}>
                    Kerrokantasi
                  </Link>
                </Navbar.Brand>
                <FormattedMessage id="navigationMenu">
                  {navigationMenu => (
                    <Navbar.Toggle aria-label={navigationMenu} aria-expanded={this.state.navbarExpanded}/>
                  )}
                </FormattedMessage>
              </Navbar.Header>
              <Navbar.Collapse>
                <ul className="nav navbar-nav">
                  {this.getNavItem('hearings', '/hearings/list')}
                  {this.getNavItem('hearingMap', '/hearings/map')}
                  {this.getNavItem('info', '/info')}
                  {this.getNavItem('ownHearings', '/user-hearings', false)}
                </ul>
              </Navbar.Collapse>
            </Navbar>
          )}
        </FormattedMessage>
      </div>
    );
  }
}

Header.propTypes = {
  // eslint-disable-next-line react/no-unused-prop-types
  dispatch: PropTypes.func,
  history: PropTypes.object,
  language: PropTypes.string,
  user: PropTypes.object,
  toggleContrast: PropTypes.func,
};

Header.contextTypes = {
  history: PropTypes.object,
};

const mapDispatchToProps = dispatch => ({
  toggleContrast: () => dispatch(toggleContrast())
});

export default withRouter(connect(state => ({
  user: getUser(state), // User dropdown requires this state
  language: state.language, // Language switch requires this state
  router: state.router, // Navigation activity requires this state
}), mapDispatchToProps)(Header));
