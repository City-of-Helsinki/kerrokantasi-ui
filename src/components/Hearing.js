import React from 'react';
import {connect} from 'react-redux';
import {push} from 'redux-router';
import Button from 'react-bootstrap/lib/Button';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';
import DeleteModal from './DeleteModal';
import {injectIntl, intlShape, FormattedMessage} from 'react-intl';
import ContactCard from './ContactCard';
import Waypoint from 'react-waypoint';

import {
  followHearing,
  postSectionComment, editSectionComment, postVote, deleteSectionComment
} from '../actions';
import SortableCommentList from './SortableCommentList';
import HearingImageList from './HearingImageList';
import LabelList from './LabelList';
import WrappedSection from './Section';
import SectionList from './SectionList';
import Sidebar from '../views/Hearing/Sidebar';
import _, {find} from 'lodash';
import Icon from '../utils/Icon';
import {
  acceptsComments,
  getClosureSection,
  getHearingURL,
  getMainSection,
  hasFullscreenMapPlugin
} from '../utils/hearing';
import {
  isSpecialSectionType,
  userCanComment,
  userCanVote
} from '../utils/section';
import getAttr from '../utils/getAttr';

export class Hearing extends React.Component {

  constructor(props) {
    super(props);

    this.state = { showDeleteModal: false, commentToDelete: { } };
  }

  openFullscreen(hearing) {
    this.props.dispatch(push(getHearingURL(hearing, {fullscreen: true})));
  }

  onPostHearingComment(text, authorName, pluginData, geojson, label, images) { // eslint-disable-line
    const {dispatch} = this.props;
    const hearingSlug = this.props.hearingSlug;
    const {authCode} = this.props.location.query;
    const mainSection = getMainSection(this.props.hearing);
    const commentData = {text, authorName, pluginData: null, authCode, geojson: null, label: null, images};
    dispatch(postSectionComment(hearingSlug, mainSection.id, commentData));
  }

  onPostSectionComment(sectionId, sectionCommentData) {
    const {dispatch} = this.props;
    const hearingSlug = this.props.hearingSlug;
    const {authCode} = this.props.location.query;
    const commentData = Object.assign({authCode}, sectionCommentData);
    dispatch(postSectionComment(hearingSlug, sectionId, commentData));
  }

  onEditSectionComment(sectionId, commentId, commentData) {
    const {dispatch} = this.props;
    const hearingSlug = this.props.hearingSlug;
    const {authCode} = this.props.location.query;
    Object.assign({authCode}, commentData);
    dispatch(editSectionComment(hearingSlug, sectionId, commentId, commentData));
  }

  onDeleteComment() {
    const {dispatch} = this.props;
    const {sectionId, commentId} = this.state.commentToDelete;
    const hearingSlug = this.props.hearingSlug;
    dispatch(deleteSectionComment(hearingSlug, sectionId, commentId));
    this.forceUpdate();
  }

  handleDeleteClick(sectionId, commentId) {
    this.setState({ commentToDelete: {sectionId, commentId}});
    this.openDeleteModal();
  }

  onVoteComment(commentId, sectionId) {
    const {dispatch} = this.props;
    const hearingSlug = this.props.hearingSlug;
    dispatch(postVote(commentId, hearingSlug, sectionId));
  }

  onFollowHearing() {
    const {dispatch} = this.props;
    const hearingSlug = this.props.hearingSlug;
    dispatch(followHearing(hearingSlug));
  }

  getFollowButton() {
    if (this.props.user === null) {
      return null;
    }
    return (
      <span className="pull-right">
        <Button bsStyle="primary" onClick={this.onFollowHearing.bind(this)}>
          <Icon name="bell-o"/> <FormattedMessage id="follow"/>
        </Button>
      </span>
    );
  }

  getClosureInfo(hearing) {
    const {formatMessage} = this.props.intl;
    const closureInfo = getClosureSection(hearing);
    if (closureInfo) {
      return closureInfo;
    }
    // Render default closure info if no custom section is specified
    return ({ type: "closure-info",
      title: "",
      abstract: "",
      images: [],
      content: formatMessage({id: 'defaultClosureInfo'}) }
    );
  }

  getLinkToFullscreen(hearing) {
    if (!hasFullscreenMapPlugin(hearing)) {
      return null;
    }
    return (
      <Button bsStyle="primary" bsSize="large" block onClick={() => this.openFullscreen(hearing)}>
        <Icon name="arrows-alt" fixedWidth/>
        <FormattedMessage id="openFullscreenMap"/>
      </Button>
    );
  }

  isSectionVotable(section, user) {
    const hearing = this.props.hearing;
    return acceptsComments(hearing) && userCanVote(user, section);
  }

  isMainSectionVotable(user) {
    const hearing = this.props.hearing;
    const section = getMainSection(hearing);
    return this.isSectionVotable(section, user);
  }

  isSectionCommentable(section, user) {
    const hearing = this.props.hearing;
    return (
      acceptsComments(hearing)
      && userCanComment(user, section)
      && !section.plugin_identifier // comment box not available for sections with plugins
    );
  }

  isMainSectionCommentable(user) {
    const hearing = this.props.hearing;
    const section = getMainSection(hearing);
    return this.isSectionCommentable(section, user);
  }

  getCommentList() {
    const {hearing, sectionComments, location, hearingSlug} = this.props;
    const mainSection = getMainSection(hearing);
    const showPluginInline = !mainSection.plugin_fullscreen;
    const user = this.props.user;
    let userIsAdmin = false;
    if (hearing && user && _.has(user, 'adminOrganizations')) {
      userIsAdmin = _.includes(user.adminOrganizations, hearing.organization);
    }
    if (!showPluginInline) {
      return null;
    }
    return (
      <div>
        <div id="hearing-comments">
          <SortableCommentList
            canVote={this.isMainSectionVotable(user)}
            displayVisualization={userIsAdmin || hearing.closed}
           section={mainSection}
           location={location}
           mainSection={mainSection}
           hearingSlug={hearingSlug}
           comments={sectionComments[mainSection.id] ?
                     sectionComments[mainSection.id].results : []}
           canComment={this.isMainSectionCommentable(hearing, user)}
           onPostComment={this.onPostHearingComment.bind(this)}
           onEditComment={this.onEditSectionComment.bind(this)}
           onDeleteComment={this.handleDeleteClick.bind(this)}
           onPostVote={this.onVoteComment.bind(this)}
           canSetNickname={user === null}
          />
        </div>
        <hr/>
      </div>
    );
  }

  openDeleteModal() {
    this.setState({ showDeleteModal: true });
  }

  closeDeleteModal() {
    this.setState({ showDeleteModal: false, commentToDelete: {} });
  }

  render() {
    const {hearing, hearingSlug, user, language, dispatch, changeCurrentlyViewed, currentlyViewed} = this.props;
    const hearingAllowsComments = acceptsComments(hearing);
    const mainSection = getMainSection(hearing);
    const showPluginInline = Boolean(!mainSection.plugin_fullscreen && mainSection.plugin_identifier);
    const closureInfoSection = this.getClosureInfo(hearing);
    const regularSections = hearing.sections.filter((section) => !isSpecialSectionType(section.type));
    const sectionGroups = groupSections(regularSections);
    return (
      <div id="hearing-wrapper">
        <LabelList className="main-labels" labels={hearing.labels}/>
        <Waypoint onEnter={() => changeCurrentlyViewed('#hearing')}/>
        <h1 className="page-title">
          {this.getFollowButton()}
          {!hearing.published ? <Icon name="eye-slash"/> : null}
          {getAttr(hearing.title, language)}
        </h1>

        <Row>
          <Sidebar
            currentlyViewed={currentlyViewed}
            hearing={hearing}
            mainSection={mainSection}
            sectionGroups={sectionGroups}
            activeLanguage={language}
            dispatch={dispatch}
            hearingSlug={hearingSlug}
          />
          <Col md={8} lg={9}>
            <div id="hearing">
              <Waypoint onEnter={() => changeCurrentlyViewed('#hearing')} topOffset={'-30%'}/>
              <div>
                <HearingImageList images={mainSection.images}/>
                <div
                  className="hearing-abstract lead"
                  dangerouslySetInnerHTML={{__html: getAttr(hearing.abstract, language)}}
                />
              </div>
              {hearing.closed ? <WrappedSection section={closureInfoSection} canComment={false}/> : null}
              {mainSection ? <WrappedSection
                showPlugin={showPluginInline}
                section={mainSection}
                canComment={this.isMainSectionCommentable(hearing, user)}
                onPostComment={this.onPostSectionComment.bind(this)}
                onPostVote={this.onVoteComment.bind(this)}
                canVote={this.isMainSectionVotable(user)}
                comments={this.props.sectionComments[mainSection.id]}
                user={user}
              /> : null}
            </div>

            {this.getLinkToFullscreen(hearing)}
            {sectionGroups.map((sectionGroup) => (
              <div id={"hearing-sectiongroup-" + sectionGroup.type} key={sectionGroup.type}>
                <Waypoint onEnter={() => changeCurrentlyViewed('#hearing-sectiongroup' + sectionGroup.name_singular)}/>
                <SectionList
                  basePath={location.pathname}
                  sections={sectionGroup.sections}
                  nComments={sectionGroup.n_comments}
                  canComment={hearingAllowsComments}
                  onPostComment={this.onPostSectionComment.bind(this)}
                  canVote={hearingAllowsComments}
                  onPostVote={this.onVoteComment.bind(this)}
                  sectionComments={this.props.sectionComments}
                  user={user}
                />
              </div>
            ))}
            <Waypoint onEnter={() => changeCurrentlyViewed('#hearing-comments')} topOffset={'-300px'}/>
            {this.getCommentList()}

            {hearing.contact_persons && hearing.contact_persons.length ?
              <h2><FormattedMessage id="contactPersons"/></h2> :
              null}
            {hearing.contact_persons && hearing.contact_persons.map((person, index) =>
              <ContactCard key={index} {...person}/> // eslint-disable-line react/no-array-index-key
            )}
          </Col>
        </Row>
        <DeleteModal
          isOpen={this.state.showDeleteModal}
          close={this.closeDeleteModal.bind(this)}
          onDeleteComment={this.onDeleteComment.bind(this)}
        />
      </div>
    );
  }
}

Hearing.propTypes = {
  intl: intlShape.isRequired,
  dispatch: React.PropTypes.func,
  hearing: React.PropTypes.object,
  hearingSlug: React.PropTypes.string,
  language: React.PropTypes.string,
  location: React.PropTypes.object,
  user: React.PropTypes.object,
  sectionComments: React.PropTypes.object,
  changeCurrentlyViewed: React.PropTypes.func,
  currentlyViewed: React.PropTypes.string
};

export function wrapHearingComponent(component, pure = true) {
  const wrappedComponent = connect((state) => ({language: state.language}), null, null, {pure})(injectIntl(component));
  // We need to re-hoist the data statics to the wrapped component due to react-intl:
  wrappedComponent.canRenderFully = component.canRenderFully;
  wrappedComponent.fetchData = component.fetchData;
  return wrappedComponent;
}

export default wrapHearingComponent(Hearing);

function groupSections(sections) {
  const sectionGroups = [];
  sections.forEach((section) => {
    const sectionGroup = find(sectionGroups, group => section.type === group.type);
    if (sectionGroup) {
      sectionGroup.sections.push(section);
      sectionGroup.n_comments += section.n_comments;
    } else {
      sectionGroups.push({
        name_singular: section.type_name_singular,
        name_plural: section.type_name_plural,
        type: section.type,
        sections: [section],
        n_comments: section.n_comments,
      });
    }
  });
  return sectionGroups;
}
