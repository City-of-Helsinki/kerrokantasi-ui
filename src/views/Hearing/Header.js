import React from 'react';
import { Col, Row, OverlayTrigger } from 'react-bootstrap';
import { injectIntl, FormattedPlural, FormattedMessage } from 'react-intl';
import FormatRelativeTime from '../../utils/FormatRelativeTime';
import LabelList from '../../components/LabelList';
import SocialBar from '../../components/SocialBar';
import Icon from '../../utils/Icon';
import getAttr from '../../utils/getAttr';

class Header extends React.Component {
  getTimetableText(hearing) { // eslint-disable-line class-methods-use-this
    if (!hearing.published) {
      return (
        <div className="timetable">
          <Icon name="clock-o"/> <del><FormatRelativeTime messagePrefix="timeOpen" timeVal={hearing.open_at}/></del> (<FormattedMessage id="draftNotPublished" />)
          <br/>
          <Icon name="clock-o"/> <del><FormatRelativeTime messagePrefix="timeClose" timeVal={hearing.close_at}/></del>
        </div>
      );
    }
    return (
      <div className="timetable">
        <Icon name="clock-o"/> <FormatRelativeTime messagePrefix="timeOpen" timeVal={hearing.open_at}/>
        <br/>
        <Icon name="clock-o"/> <FormatRelativeTime messagePrefix="timeClose" timeVal={hearing.close_at}/>
      </div>
    );
  }

  render() {
    const { hearing, activeLanguage, reportUrl, eyeTooltip } = this.props;
    return (
      <div className="hearing-header well">
        <h1>
          {!hearing.published
            ? <OverlayTrigger placement="bottom" overlay={eyeTooltip}><Icon name="eye-slash"/></OverlayTrigger>
            : null}
          {' '}
          {getAttr(hearing.title, activeLanguage)}
        </h1>
        <Row className="hearing-meta">
          <Col xs={12}>
            <LabelList className="main-labels" labels={hearing.labels} />
          </Col>
          <Col xs={12} sm={6}>
            {this.getTimetableText(hearing)}
          </Col>
          <Col xs={12} sm={6}>
            <div className="comment-summary">
              {hearing.n_comments
                ? <div className="commentNumber">
                  <Icon name="comment-o" /> {' '}
                  <FormattedPlural
                      value={hearing.n_comments}
                      one={<FormattedMessage id="totalSubmittedComment" values={{ n: hearing.n_comments }} />}
                      other={<FormattedMessage id="totalSubmittedComments" values={{ n: hearing.n_comments }} />}
                  />
                </div>
              : null}
              {reportUrl
                ? <div className="report-download">
                  <a href={reportUrl}>
                    <small>
                      <Icon name="download" /> <FormattedMessage id="downloadReport" />
                    </small>
                  </a>
                </div>
              : null}
            </div>
          </Col>
        </Row>
        <SocialBar />
      </div>
    );
  }
}

Header.propTypes = {
  eyeTooltip: React.PropTypes.element,
  hearing: React.PropTypes.object,
  reportUrl: React.PropTypes.string,
  activeLanguage: React.PropTypes.string,
};

export default injectIntl(Header);
