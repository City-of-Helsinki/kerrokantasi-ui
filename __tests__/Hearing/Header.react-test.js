import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {HeaderComponent} from '../../src/components/Hearing/Header';
import {mockStore, getIntlAsProp} from '../../test-utils';

// Renders the Hearings component using enzymes shallow rendering.
// You can pass props you want to override as a parameter.
const setup = propOverrides => {
  const {mockHearingWithSections} = mockStore;
  const props = Object.assign({
    hearing: mockHearingWithSections.data,
    sections: mockHearingWithSections.data.sections,
    activeLanguage: 'fi',
    dispatch: () => {},
    location: {
      pathname: mockHearingWithSections.data.sections[0].id
    },
    match: {
      params: {
        sectionId: mockHearingWithSections.data.sections[0].id
      }
    },
  }, propOverrides);

  const wrapper = shallow(<HeaderComponent intl={getIntlAsProp()} {...props} />);

  return {
    props,
    wrapper
  };
};

test('Header component should render as expected', () => {
  const {wrapper} = setup();
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});
