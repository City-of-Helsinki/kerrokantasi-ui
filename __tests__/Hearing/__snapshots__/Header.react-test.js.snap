// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render as expected 1`] = `
<Fragment>
  <div
    className="header-section"
  >
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={false}
    >
      <div
        className="hearing-header"
      >
        <Row
          bsClass="row"
          componentClass="div"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
          >
            <h1
              className="hearing-header-title"
            >
              Ideoi Isosaaren tulevaisuutta
            </h1>
          </Col>
        </Row>
        <Row
          bsClass="row"
          componentClass="div"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
          >
            <div
              className="header-abstract lead"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "Tämä on suomenkielinen testiabstrakti. Sisältää huikean kuvauksen kyseisestä kuulemisesta.",
                }
              }
            />
          </Col>
        </Row>
        <div
          className="hearing-meta"
        >
          <div
            className="hearing-meta__element timetable"
          >
            <Icon
              Component="span"
              name="clock-o"
            />
            <span
              className="timetable-texts"
            >
              <FormatRelativeTime
                messagePrefix="timeOpen"
                timeVal="2017-10-22T21:00:00Z"
              />
              <br />
              <FormatRelativeTime
                formatDate={[Function]}
                formatTime={[Function]}
                messagePrefix="timeClose"
                timeVal="2200-11-28T22:00:00Z"
              />
            </span>
          </div>
          <div
            className="hearing-meta__element comment-summary"
          >
            <Icon
              Component="span"
              name="comment-o"
            />
            <span
              className="comment-summary-texts"
            >
              <div
                className="commentNumber"
              >
                <FormattedPlural
                  one={
                    <FormattedMessage
                      id="totalSubmittedComment"
                      values={
                        Object {
                          "n": 1,
                        }
                      }
                    />
                  }
                  other={
                    <FormattedMessage
                      id="totalSubmittedComments"
                      values={
                        Object {
                          "n": 1,
                        }
                      }
                    />
                  }
                  style="cardinal"
                  value={1}
                />
              </div>
              <div>
                <AnchorLink
                  href="#comments-section"
                  offset="100"
                >
                  <FormattedMessage
                    id="headerWriteCommentLink"
                    values={Object {}}
                  />
                </AnchorLink>
              </div>
            </span>
          </div>
          <div
            className="hearing-meta__element language-select"
          >
            <Icon
              Component="span"
              className="user-nav-icon"
              name="globe"
            />
            <span
              className="language-select__texts"
            >
              <div
                className="translation-not-available"
              >
                <FormattedMessage
                  id="hearingTranslationNotAvailable"
                  values={Object {}}
                />
              </div>
              Kuuleminen saatavilla vain
               
              <withRouter(Connect(LinkWithLangComponent))
                className="language-select__language"
                to={
                  Object {
                    "path": "7CxRxrlvrW9uCCjEaB9rW1dENHQMudeK",
                    "search": "?lang=fi",
                  }
                }
              >
                suomeksi
              </withRouter(Connect(LinkWithLangComponent))>
            </span>
          </div>
        </div>
      </div>
    </Grid>
  </div>
</Fragment>
`;
